src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1528,7 +1528,7 @@ private IEnumerable<TypeInferenceInfo> GetTypesForRecursivePattern(RecursivePatt
                                 return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
                             }
 
-                            elementTypesBuilder.Add(patternType.InferredType.WithoutNullability());
+                            elementTypesBuilder.Add(patternType.InferredType);
                         }
 
                         var type = Compilation.CreateTupleTypeSymbol(
```

Comment:
❔ Do we need to pass explicit nullabilities here? I thought I remembered seeing a reference to a compiler API bug related to this.

Reply:
Good catch: so, the code as was previously written simply wasn't processing top-level nullability in the first place. So the code as you reviewed it was "correct" in that it wasn't introducing more bugs, but the bug was already. I've fixed the bug, added a test, and am now applying the workaround at the compiler layer again.